b1ddi/resource_ipam_subnet.go

```diff
@@ -675,3 +661,25 @@ func resourceIpamsvcSubnetDelete(ctx context.Context, d *schema.ResourceData, m
 
 	return nil
 }
+
+func generateAddressPath(d *schema.ResourceData, path string) (string, diag.Diagnostics) {
```

Comment:
most of the validation in this func could have been done in the `ValidateFunc` in the schema definition. 

This should have only been a simple check whether next_avaible_id is empty or not empty, return string accordingly.

Reply:
To see if an attrribute value has been provided by user, you can also use `d.HasChange` instead of checking with empty strings.

Reply:
What if the user passes an empty string? In this case, `HasChange` would be true right?

Reply:
The validation for empty string and if it contains /ipam etc should be done in `ValidateDiags` function, so that the error can be caught during `terraform plan` itself and not during `terraform apply` an example would be 

`ValidateFunc: validation.StringMatch(regexp.MustCompile(`^ipam\/(range|subnet|address_block)\/[0-9a-f-].*$`), ""),

You can check other validation functions that maybe useful as well here for e.g if not empty - https://github.com/hashicorp/terraform-plugin-sdk/blob/ffbf0104398c0aa91aa3a82aff4b67e260677454/helper/validation/strings.go#L17C7-L17C7